To-do 리스트 Redux를 사용하여 리팩터링 하기

psy082 opened the pull request:
재택이 끝나고 사무실 출근을 시작하면서 과제에 충분히 시간을 쓰지 못하게 되었는데 조금씩 더 많이 시간을 사용하려고 합니다.
아직은 redux, react-redux만 설치했기 때문에 리뷰 받을 내용이 없는데
조금씩 작성해보도록 하겠습니다!

psy082 committed:
init: 프로젝트 초기 설정

- redux, react-redux 설치

moonkii commented:
![image](https://user-images.githubusercontent.com/46776329/131680133-57541a16-2247-4236-89da-90de328f9924.png)

첫번 째 과제는 강의 내용을 복습해보는 것에 가까워요.
기본적인 Redux 사용법과 코드 패턴을 익히기 위한 연습이라 생각하고 해보세요 ;)

psy082 committed:
refactor: App component 상태 변경 로직 분리

psy082 committed:
feat: redux store 생성

- 빈 store 생성
- react-redux Provider로 store와 react 연결

psy082 committed:
refactor: 상태 관리 로직 store로 이동

- App component에 있는 상태 관리 로직 store로 이동
- App component는 redux의 dispatch와 selector를 사용하도록 변경

psy082 committed:
refactor: App component에 있는 action creator 함수들 분리

psy082 committed:
test: App component test mock Provider 작성

psy082 committed:
refactor: store에서 reducer 분리

psy082 committed:
test: reducer test 작성

- updateTaskTitle action test 작성
- addTask action test 작성
- deleteTask action test 작성

moonkii commented on src/store/reducer.js:
```diff
@@ -0,0 +1,36 @@
+
+export default function reducer(state = initialState, action) {
+  const { newId, taskTitle, tasks } = state;
+  switch (action.type) {
```
`switch` 문은 코드 가독성이 좋지 않아요.
그래서 강의를 보면 차라리 if 문을 나눠서 여러번 쓰는걸 볼 수 있으실 거에요.
리팩터링을 할 때는 기존의 어떤 점이 불편하거나 문제가 있었고, 변경한 부분은 어떤 점에서 더 좋은지 비교하면서 고민하시면 좋아요.

psy082 replied:
혹시 swtich가 가독성이 떨어지는 이유가 있을까요?

언듯 생각하면 action.type 값에 대한 병렬 처리라서 switch를 쓰는 것도 좋을 것 같은데
궁금해서 여쭤봅니다!

moonkii replied:
과제의 제한조건에도 보면 switch 를 사용하지 않는다고 명시되어있어요.
기본적으로 코드숨에서는 코드가독성을 헤치는 문법은 지양하고 있어요.
과제를 진행할 때 꼭 전체적으로 요구사항을 잘 숙지하고 진행해주세요.

먼저 지금 같은 구조는 switch 가 case 가 증가함에 따라 길이가 길어지고, depth도 깊어져서 코드 가독성이 떨어져요.
if ~ else 도 마찬가지죠. 강의에서도 더 리펙터링을 할 수 있지만 일단은 if ~ else 를 쓰지 않고 각각 if 로 작성한 이유가 여기에 있어요.

switch를 하나의 함수로 본다면 switch 내부에서 너무 많은 것들이 처리되고 있는걸 알 수 있을거에요.

psy082 replied:
아하 넵넵 감사합니다. swtich가 가독성이 떨어지는 이유가 언듯 바로 이해되지 않아서 질문해봤습니다!

psy082 committed:
test: reducer test 추가

- initial state일 때의 값 확인 test
- 정의되지 않은 action을 호출했을 때의 처리 test

moonkii commented:
테스트를 확인해보시고 테스트 커버리지가 채워지지 않은 부분도 마저 채워보세요 ;)

psy082 committed:
refactor: 각 컴포넌트 폴더로 분리

psy082 committed:
refactor: ListContainer 분리

- list 상태 로직 ListContainer로 이동
- Page 인터페이스 변경
- Appdptj Page 호출부분 변경

psy082 committed:
test: ListContainer test 작성, Page test 수정

psy082 committed:
test: ListContainer test 수정

- 모든 task 문구에 대한 test 추가

psy082 committed:
refactor: InputContainer 분리

- taskTitle 상태와 관련된 로직 이동
- InputContiner를 위한 test 작성
- Page, Page test 삭제

moonkii commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,105 @@
+    });
+  });
+
+  describe('call action not pre-defined', () => {
```
describe 는 테스트의 서술 대상이에요. 여기서 테스트하는 주체는 reducer 이기 때문에 context 로 맥락을 나눠줄 수는 있어도 describe 를 작성하는건 불필요핼 것 같아요.

psy082 replied:
아하 이해 했습니다. 테스트 코드 명에서 descibe가 테스트 서술 대상이 아닌 부분을 전부 수정했습니다!

moonkii commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,105 @@
+
+describe('reducer', () => {
+  describe('initial state', () => {
+    it('uses initial state and update taskTitle', () => {
```
uses 라는 표현보다는 returns 같은 표현이 어떨까요? 🤔

psy082 replied:
넵넵 이 부분은 아예 initial state를 사용하는 context와 새로운 state를 사용하는 context로 변경해서 작성해봤습니다!

moonkii commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,105 @@
+} from './actions';
+
+describe('reducer', () => {
+  describe('initial state', () => {
```
describe 를 확인해보고 싶을 때는 it 에다가 describe 대상을 넣어보시면 돼요.
initial state uses initial state and update taskTitle 이라고 하면 어떻게 해석이되나요?

psy082 replied:
완전 이상합니다. describe와 it이 어떻게 서술되어야 하는 지 알았습니다. 이 부분도 일괄 수정했습니다!

psy082 committed:
refactor: reducer action 처리 로직 switch에서 if문으로 수정

psy082 committed:
test: InputContainer test 수정

- redux state에 저장된 첫 렌더링 시 taskTitle 값이 input에 표시되는 지 test
- input 값 입력 시 updateTaskTitle action이 호출되는 지 test
- 추가 버튼 클릭시 addTask action이 호출되는 지 test

psy082 committed:
InputContainer test 수정

- 테스트 명 수정

psy082 committed:
ListContainer test 수정

- 테스트 명 수정
- 완료 버튼을 눌렀을 때의 deleteTask action 호출 테스트 추가

psy082 committed:
test: reducer test 수정

- initial state 사용하는 경우 테스트를 updateTaskTitle 내부로 이동
- 테스트 명 수정

moonkii commented on src/Container/Input/InputContainer.test.jsx:
```diff
@@ -6,23 +6,53 @@ import InputContainer from './InputContainer';
-  useSelector.mockImplementation((selector) => selector({
-    taskTitle: 'New Title',
-  }));
+  it('has taskTitle of redux state as a input value', () => {
+    useSelector.mockImplementation((selector) => selector({
+      taskTitle: 'New Title',
+    }));
```
이 부분은 `useSelector` 를 mock 해주는 부분이에요!
테스트를 위해 미리 환경을 세팅하는 부분이기 때문에 describe 내부에 존재해야하고, it 으로 감싸주는건 적절하지 않아요. 
이렇게 코드를 작성하면 매번 it 마다 `useSelector`를 mock 해줘야할거에요.

moonkii commented on src/Container/Input/InputContainer.test.jsx:
```diff
@@ -6,23 +6,53 @@ import InputContainer from './InputContainer';
+    });
+  });
+
+  it('calls addTask action when add button is clicked', () => {
```
when 이하는 생략해주셔도 괜찮을것 같아요.
~했을 때 라는건 맥락이에요. 그리고 맥락(context)는 명시해줄 필요가 있을 때만 해주는게좋아요.

moonkii commented:
테스트 코드를 작성할 때 각 테스트 코드가 어떤 역할을 하는지 하나하나 공부해보시면 좋아요.

moonkii commented on src/Container/List/ListContainer.test.jsx:
```diff
@@ -0,0 +1,51 @@
+  });
+
+  it('calls deleteTask action when complete button is clicked', () => {
+    const dispatch = jest.fn();
+
+    useDispatch.mockImplementation(() => dispatch);
```
이 부분은 describe 내부로 빼주시면 좋아요.
그리고 beforeEach 에서 mock 함수를 초기화해주는게 안전해요.

moonkii commented on src/Container/List/ListContainer.test.jsx:
```diff
@@ -0,0 +1,51 @@
+
+    useDispatch.mockImplementation(() => dispatch);
+
+    useSelector.mockImplementation((selector) => selector({
+      tasks: [
+        { id: 1, title: '아무것도 하지 않기 #1' },
+        { id: 2, title: '아무것도 하지 않기 #2' },
+      ],
+    }));
```
테스트마다 차이가 없다면 이부분도 마찬가지로 describe 내부에 선언해주시면 좋아요.
